### PR TITLE
Enable client to use multiple connections to h2 proxy server

### PR DESCRIPTION
**Pull request by darhwa**
The current one-h2-tunnel-for-all model can only consume a proportion of ISP-provided bandwidth because of QoS. My trivial solution is to set up more than one outgoing h2 connections, to leverage more bandwidth while retaining good latency.

Currently I only implemented round-robin policy, and hardcoded the upper limit of connections to be 10.

### CODE REVIEW

**Comment by klzgrad** [comment]
How much more bandwidth? I need numbers.

**Comment by klzgrad** [comment]
It was closed because of the rebase. Discussion can continue.

**Comment by darhwa** [comment]
Some download tests through naiveproxy:
single thread:
  aria2c https://lax-ca-us-ping.vultr.com/vultr.com.1000MB.bin  -->  6.3MiB/s
multithread (original naiveproxy):
  aria2c -x6 https://lax-ca-us-ping.vultr.com/vultr.com.1000MB.bin  -->  6.3MiB/s
multithread (naiveproxy using --concurrent=6):
  aria2c -x6 https://lax-ca-us-ping.vultr.com/vultr.com.1000MB.bin  -->  34MiB/s

My ISP is China Mobile 500M.

**Comment by klzgrad** [comment]
What about 2?

**Comment by klzgrad** [comment]
To let you know what I'm getting at, first I'd say I'm open to ideas that improve performance. But this is already an anti-security feature. One connection multiplexing is better at traffic obfuscation. With more connections it's easier to recognize patterns in each one. Even if this is to be used it's definitely not going to be as high as 10, which is a clear violation of usual browser limits. If 2, 3 connections can work that would be an acceptable number.

Second, I suspect this number you got is an artificial one, which won't be replicated with normal websites. Maybe it's an artifact of how this demo server is not aggressive enough to increase TCP recv windows.

Third, even if this is replicated with normal websites there's more to try than simply increasing the number of TCP connections. The point of TCP congestion control is to find out the available bandwidth in the current path, and multiplying connection number is only one way to tune congestion control. There are other ways. So I would try those first. Or maybe QUIC?

Lastly, this is architecturally tied to https://github.com/klzgrad/naiveproxy/issues/61, and the solution is not necessarily one that directly modifies the net stack. The solution can be reconsidered.

In short, more consideration is needed.

**Comment by klzgrad** [comment]
I tested it. 2 is enough. 3 is overkill. But the effect is real that a single connection is throttled at certain level.

I think I'll allow this feature in general, but it will be marked experimental and not advertised in the doc. Also the maximum connection number will be lowered  to 4 or something.

I may use a slightly different implementation following the general idea in your patch.

**Comment by klzgrad** [comment]
Since I'm making a different implementation, some comments about your previous patch:
- You shouldn't mix in unrelated changes even if it looks like a bug fix. Make another patch or report a bug.
- C++17 is not supported in Chromium coding style.
- I'm not comfortable with that naked ints being potentially used across threads. Use atomics or even better locate a better API level.
- concurrent -> concurrency

**Comment by darhwa** [comment]
> Third, even if this is replicated with normal websites there's more to try than simply increasing the number of TCP connections. The point of TCP congestion control is to find out the available bandwidth in the current path, and multiplying connection number is only one way to tune congestion control. There are other ways. So I would try those first. Or maybe QUIC?

Thanks for this hint. I will try.

**Comment by klzgrad** [comment]
I added this option in 68fcd2938a9eb3ef5a8472d65ef83c3b8addf0bb.

I see in my tests that when real congestion happens this isn't going to help. It's only when the network is idle and your single connection is throttled at 10MB/s it'd help with multiple connections.
